src/frequenz/sdk/timeseries/battery_pool/battery_pool.py

```diff
@@ -0,0 +1,142 @@
+from ...microgrid.component import ComponentCategory
+from ._methods import BaseMethod, ResamplerArgs, SendOnUpdate
+from ._metric_calculator import PowerBoundsCalculator, SoC_Capacity_Calculator
+from ._result_types import PowerBounds, SoC_Capacity
+
+
+class BatteryPool:
+    """Calculate high level metrics for a pool of the batteries.
+
+    BatterPool provided methods for fetching different battery and inverter values
+    from microgrid and aggregate then in a high level metrics.
+    Depending from the metrics the results are send in two ways:
+        * when the metric change
+        * with the resampling frequency (specified in the resampler actor outside)
+    """
+
+    def __init__(
+        self,
+        batteries_status_receiver: Receiver[BatteryStatus],
+        min_update_interval: timedelta,
+        resampler_args: ResamplerArgs,
+        batteries_id: Optional[Set[int]] = None,
+    ) -> None:
+        """Create the class instance
+
+        Args:
+            batteries_status_receiver: Receiver to receive status of the batteries.
+                Receiver maxsize has to be 1 to fetch only the latest status.
+            min_update_interval: Some metrics are send only when they change. For these
+                metrics min_update_interval is the minimum time interval between the
+                following messages.
+            resampler_args: Arguments needed for sending metric request to the.
+                resampling actor
+            batteries_id: What batteries should be included in the battery pool.
+                If None or empty, then all batteries from the microgrid will be used.
+        """
+        if batteries_id:
+            self._batteries = batteries_id
+        else:
+            self._batteries = self._get_all_batteries()
+
+        self._working_batteries: Set[int] = set()
+
+        self._update_battery_status_task = asyncio.create_task(
+            self._update_battery_status(batteries_status_receiver)
+        )
+        self._min_update_interval = min_update_interval
+        self._resampler_args = resampler_args
+
+        self._active_methods: Dict[str, BaseMethod[Any]] = {}
+
+    def _get_all_batteries(self) -> Set[int]:
+        """Get all batteries from the microgrid
+
+        Returns:
+            All batteries in the microgrid.
+        """
+        graph = microgrid.get().component_graph
+        return {
+            battery.component_id
+            for battery in graph.components(
+                component_category={ComponentCategory.BATTERY}
+            )
+        }
+
+    async def _update_battery_status(self, receiver: Receiver[BatteryStatus]):
+        async for status in receiver:
+            self._working_batteries = status.get_working_batteries(self._batteries)
+            for item in self._active_methods.values():
+                item.update_working_batteries(self._working_batteries)
+
+    async def soc_capacity(
```

Comment:
This goes in general for the common interface for stuff returning new receivers (i.e. logical meter too). I think we should be more explicit about a **new** receiver being created, otherwise people might think they can just call `soc_capacity()` whenever they need a receiver. It will make the method names a bit uglier, but I think is more clear.

* `p.new_soc_capacity_receiver()`?
* Maybe return a channel(or wrapper) instead, so users do `p.soc_capacity.new_receiver()`?
* `p.new_receiver(SOC_CAPACITY)`?

Can be left for a future PR, but I would like to address it soon, before we keep adding methods like these.

/cc @matthias-wende-frequenz

Reply:
I was thinking that with approach 2 we could add different ways to retrieve the data:

```py
current = p.soc_capacity.current
resampled_receiver = p.soc_capacity.new_receiver()
change_receiver = p.soc_capacity.new_changes_receiver()
```

This would also remove the inconsistent interface we have right now, where `soc_capacity()` could be sending messages to the channel on every resample period, or on change, is not explicit, and it would be inconsistent for different metrics, which makes me uneasy.

We could also have `new_receiver(only_changes=True)` for example, but if we don't want to support getting just a resampled receiver, we need to assert `only_changes` is `True` at runtime, and we can´t use the type system to help us find this at "compile-time".

Reply:
https://github.com/frequenz-floss/frequenz-sdk-python/discussions/215